### PR TITLE
UCP/UCT/API: another client-server connection establishment flow

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
This PR add new UCP client-server connection establishment flow. The main difference from origin is possibility to handle connection requests on listener created on one worker but create an endpoint ("accept") on other worker. Also added `ucp_listener_reject` function to avoid memory leaks and all related UCT and testing things.

Please not, this PR is created against topic branch (not master) to make possible overview of the feature from top to bottom.

@alinask @yosefe pls take a look

### CODE REVIEW

**Comment by shamisp** [comment]
Adding @gmegan

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7167/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7170/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7171/ for details (Mellanox internal link).

File: src/ucp/core/ucp_listener.c [hunk 1]
```diff
@@ -215,11 +238,22 @@ ucs_status_t ucp_listener_create(ucp_worker_h worker,
...
-            UCP_CHECK_PARAM_NON_NULL(params->accept_handler.cb, status, goto err_free);
-            listener->cb  = params->accept_handler.cb;
-            listener->arg = params->accept_handler.arg;
+            UCP_CHECK_PARAM_NON_NULL(params->accept_handler.cb, status,
+                                     goto err_free);
+            listener->cb             = params->accept_handler.cb;
+            listener->accept_addr_cb = NULL;
+            listener->arg            = params->accept_handler.arg;
+        } else if (params->field_mask &
+                   UCP_LISTENER_PARAM_FIELD_ACCEPT_ADDR_HANDLER) {
```
Comment (alinask) [inline src/ucp/core/ucp_listener.c]: do we need to prevent a case here where both flags are set?

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -1664,6 +1685,21 @@ ucs_status_ptr_t ucp_ep_modify_nb(ucp_ep_h ep, const ucp_ep_params_t *params);
...
+/**
+ * @ingroup UCP_WORKER
+ *
+ * @brief Reject @ref ucp_ep_h "endpoint" creation.
+ *
+ * This routine releases the @ref ucp_ep_address_h and notifies client if there
+ * was set @ref ucp_ep_params_t::err_handler with status @ref UCS_ERR_CANCELED.
+ *
+ * @param [in]  listener    Handle to the listener which created the address.
+ * @param [in]  ep_addr     Handle to the endpoint address to reject.
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: This is an address (not ep) which can be used for ep creation or rejecting.

File: src/ucp/api/ucp_def.h [hunk 3]
```diff
@@ -341,6 +352,23 @@ typedef struct ucp_err_handler {
...
 
+/**
+ * @ingroup UCP_WORKER
+ * @brief A callback for handling of incoming connection request with a client
+ * endpoint address which is connecting to a listener @ref ucp_listener_h.
+ *
+ *  This callback routine is invoked on the server side upon a remote client
+ *  creates an endpoint. The user can pass an argument to this callback.
+ *  The user is responsible for releasing the @a ep_addr handle using the
+ *  @ref ucp_ep_create or @ref ucp_ep_reject routine.
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/api/ucp_def.h]: ep is not exists yet, so create or reject

File: src/ucp/core/ucp_ep.h [hunk 4]
```diff
@@ -301,6 +301,24 @@ typedef struct {
...
+                                                  full or partial */
+    /* packed worker address follows */
+} UCS_S_PACKED ucp_wireup_sockaddr_priv_t;
+
+
+typedef struct ucp_ep_address {
+    struct ucp_listener         *listener;
+    void                        *conn_req_id;
+    ucp_wireup_sockaddr_priv_t  priv_addr;
+    /* packed worker address follows */
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.h]: i think yes, the address will be corrupted if some fields will be placed here

File: src/ucp/api/ucp_def.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (alinask) [reply src/ucp/api/ucp_def.h]: you wrote _releasing_ with *ep_create*

File: src/ucp/core/ucp_listener.c [hunk 5]
```diff
@@ -90,63 +104,81 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
...
 
-    if (client_data->is_full_addr) {
-        /* create endpoint to the worker address we got in the private data */
-        status = ucp_ep_create_to_worker_addr(listener->wiface.worker, &params,
-                                              &remote_address, UCP_EP_CREATE_AM_LANE,
-                                              "listener", &ep);
-        if (status != UCS_OK) {
-            goto err_free_address;
+    if (listener->cb) {
+        status = ucp_listener_ep_create(listener, client_data, accept);
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_listener.c]: Good question... Since we have 2 types of accept callbacks and the new one does not create EP until user calls `ucp_ep_create`. We don't need to initiate wireup because we don't know on which worker. On other hand, user can set `UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER` flag and pass `NULL`, let's say, to ignore incoming EPs. I think this is not valid usage because it will cause memory leak, user should call `ucp_ep_destroy` for created EPs on server.

File: src/ucp/api/ucp_def.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/api/ucp_def.h]: Yes, address handle will be released during the ep creation. it can't be usable for > 1 time.

File: src/ucp/core/ucp_ep.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.h]: ucp_ep_address_t also can't grow at this place.

File: src/ucp/core/ucp_listener.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (alinask) [reply src/ucp/core/ucp_listener.c]: but for the current flow - before the changes in this pr, the wireup would start even if there is no listener's callback, why should this change? (if neither UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER nor the new flag is set)

[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_listener.c]: ep is connected to ep for client-server flow. Even for the current flow, UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER should be mandatory to avoid memory leak related to EP creation on server side. User should handle connection request and destroy the EP after communications.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7178/ for details (Mellanox internal link).

File: src/ucp/core/ucp_listener.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (alinask) [reply src/ucp/core/ucp_listener.c]: It is not mentioned in the api that this flag is mandatory. Also, if this flag won't be set, in the "worst" case, the ep on the server side would be destoyed as part of the worker's destroy flow.

File: src/ucp/api/ucp.h [hunk 6]
```diff
@@ -889,6 +904,12 @@ typedef struct ucp_ep_params {
      */
     ucs_sock_addr_t         sockaddr;
 
+    /**
+     * Address of a remote endpoint directly connect to.
```
Comment (alinask) [inline src/ucp/api/ucp.h]: 1. I think that need to clarify the wording here.
2. Does this field come with a corresponding bit in field_mask?
3. Is ep_addr used only when calling ucp_ep_create from the server side?

File: src/ucp/core/ucp_listener.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_listener.c]: > the ep on the server side would be destoyed as part of the worker's destroy flow

For long living server application this is a memory leak. But anyway I have to return initial behavior for backward compatibility. So, if user does not setup any callback, the server will create internal EP, which can be destroyed only during worker destroy (with current API)

[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_listener.c]: Fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7183/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7185/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 7]
```diff
@@ -1664,6 +1702,21 @@ ucs_status_ptr_t ucp_ep_modify_nb(ucp_ep_h ep, const ucp_ep_params_t *params);
 ucs_status_ptr_t ucp_ep_close_nb(ucp_ep_h ep, unsigned mode);
 
 
+/**
+ * @ingroup UCP_WORKER
+ *
+ * @brief Reject @ref ucp_ep_h "endpoint" creation.
+ *
+ * This routine releases the @ref ucp_ep_address_h and notifies client if there
+ * was set @ref ucp_ep_params_t::err_handler with status @ref UCS_ERR_CANCELED.
```
Comment (evgeny-leksikov) [inline src/ucp/api/ucp.h]: TODO: add new status code UCS_ERR_REJECTED

File: src/ucp/core/ucp_listener.c [hunk 8]
```diff
@@ -90,63 +104,83 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
...
     }
 
-    if (client_data->is_full_addr) {
-        /* create endpoint to the worker address we got in the private data */
-        status = ucp_ep_create_to_worker_addr(listener->wiface.worker, &params,
-                                              &remote_address, UCP_EP_CREATE_AM_LANE,
-                                              "listener", &ep);
-        if (status != UCS_OK) {
-            goto err_free_address;
+    if (listener->cb) {
```
Comment (alinask) [inline src/ucp/core/ucp_listener.c]: shouldn't the rest of this function be thread safe to protect the worker?

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_listener.c]: Seems like yes because 

1. 
 https://github.com/openucx/ucx/blob/706ffc57d5587e2c012a6feced2a5aec36171edb/src/ucp/core/ucp_listener.c#L230

2. 
 https://github.com/openucx/ucx/blob/706ffc57d5587e2c012a6feced2a5aec36171edb/src/uct/ib/rdmacm/rdmacm_iface.c#L140

and this is not a regression, right?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7187/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7189/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@alinask pls review again

File: src/ucp/core/ucp_listener.c [hunk 9]
```diff
@@ -90,83 +104,99 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
...
-        /* the listener's ep should be aware of the sent address from the client */
-        ep->flags |= UCP_EP_FLAG_LISTENER;
+    } else {
+        ucs_assert((listener->cb == NULL) && (listener->addr_cb == NULL));
+        /* No callback, but create internal EP for backward compatibility */
+        status = ucp_listener_ep_create(listener, client_data, accept);
     }
 
-    ucp_ep_update_dest_ep_ptr(ep, client_data->ep_ptr);
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&worker->mt_lock);
```
Comment (alinask) [inline src/ucp/core/ucp_listener.c]: dont need it in the end of the function?

File: src/ucp/core/ucp_ep.c [hunk 10]
```diff
@@ -70,14 +70,16 @@ ucs_status_t ucp_ep_new(ucp_worker_h worker, const char *peer_name,
...
-    ucp_ep_ext_gen(ep)->user_data   = NULL;
-    ucp_ep_ext_gen(ep)->err_cb      = NULL;
+    ep->worker                             = worker;
+    ep->cfg_index                          = ucp_worker_get_ep_config(worker,
+                                                                      &key);
+    ep->am_lane                            = UCP_NULL_LANE;
+    ep->flags                              = 0;
+    ep->conn_sn                            = -1;
+    ucp_ep_ext_gen(ep)->listener           = NULL;
+    ucp_ep_ext_gen(ep)->user_data          =
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c]: It was on the same line with `err_cb` which is initialized with `NULL`, will fix it for better readability

File: src/ucp/core/ucp_ep.c [hunk 11]
```diff
@@ -378,6 +380,75 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
...
+            goto free_address;
+        }
+    } else {
+        status = ucp_ep_create_sockaddr_aux(worker, &params, &remote_address,
+                                            ep_p);
+        if (status != UCS_OK) {
+            goto free_address;
+        }
+
+        /* the listener's ep should be aware of the sent address from the client */
```
[the remark this answers is not in the document]
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c]: Ok

File: src/ucp/core/ucp_listener.c (hunk 9 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_listener.c]: `uct_worker_progress_register_safe` and `ucp_worker_signal_internal` are thread-safe

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7190/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Please add justification why are doing this.

**Comment by evgeny-leksikov** [comment]
@shamisp it's in description:
> The main difference from origin (flow) is possibility to handle connection requests on listener created on one worker but create an endpoint ("accept") on other worker.

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by miked-mellanox** [comment]
@shamisp maybe you can create pr template which will have fields required for what,why,how. Github seems support it: https://blog.github.com/2016-02-17-issue-and-pull-request-templates/

**Comment by shamisp** [comment]
I was thinking about this. I didn't want to be super formal on this but
just including very basic justification why are doing this. What is the use
case, etc.
I will make it up.

**Comment by shamisp** [comment]
@miked-mellanox #2689

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7198/ for details (Mellanox internal link).

**Comment by gmegan** [comment]
If I understand this correctly, there are two possible use cases that I might have for this, but I am trying to understand how I would use this for those cases. Both cases involve connecting two groups of processes together, say two groups of MPI processes or two groups of shmem processes.

In the first case, I want to connect two groups of processes, G0 and G1. But I want to do this programmatically only using one process from each group. So say P0 from G0 connects to listener on P1 from G1, but instead gets back a set of eps, one for each process in G1. I think this method would let me make a listener on P1 with an ep for any one worker in G1 that I have an address for. What I really want is a listener that creates many eps on connect.

In the second case, I want to load balance work over a group of processes by connecting to a single process in the group and get back an ep to some random process worker in that group. I could already do this with the current API on the ucp client side by randomly selecting some address out of a predefined set and connecting to it to get a randomly assigned ep. But it would be a better load balancer design if I can do the randomization on the server side and the client only knows one address to connect, and nothing about how many peers are behind that address.

I am trying to understand if/how this change would allow either of those cases.

**Comment by evgeny-leksikov** [comment]
@gmegan Initially, the feature was designed for client-server apps which have many (UCP) workers per 1 (UCP) listener on server side. For example, 1 (or several) listeners with well known address and pool of threads/workers. In this case client creates EP (connection) to the socket address but server is able to rebalance this connection request across the pool creating the backward EP (like "accept") on any available worker.
But it's great if you see use cases for MPI/SHMEM as well.

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 12]
```diff
@@ -229,6 +304,32 @@ class test_ucp_sockaddr : public ucp_test {
         }
     }
 
+    void wait_for_server_reject(bool wakeup)
+    {
```
Comment (alinask) [inline test/gtest/ucp/test_ucp_sockaddr.cc]: what does reject on the server side mean? when would it happen?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 13]
```diff
@@ -229,6 +304,32 @@ class test_ucp_sockaddr : public ucp_test {
...
+        ucs_time_t time_limit = ucs_get_time() +
+                                ucs_time_from_sec(UCP_TEST_TIMEOUT_IN_SEC);
+
+        while ((sender().get_rejected_cntr() == 0) &&
+               (ucs_get_time() < time_limit)) {
+            check_events(sender().worker(), receiver().worker(), wakeup, NULL);
+        }
+        EXPECT_GT(time_limit, ucs_get_time());
+        EXPECT_EQ(1ul, sender().get_rejected_cntr());
```
Comment (alinask) [inline test/gtest/ucp/test_ucp_sockaddr.cc]: what should trigger the reject here?

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 12 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: `ucp_test_base::entity::reject_ep_addr_cb` is called

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 13 above)
Comment (evgeny-leksikov) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: [Err callback](https://github.com/openucx/ucx/pull/2672/files#diff-c19a05052b898dabf25801d90564171bR418) with status `UCS_ERR_REJECTED` increments the counter

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 12 above)
Comment (alinask) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: so wait_for_server_reject means that the server will _do_ the reject,
and wait_for_client_reject means that the client will _receive_ the reject from the server?

File: src/ucp/api/ucp.h [hunk 14]
```diff
@@ -844,11 +859,14 @@ typedef struct ucp_ep_params {
     uint64_t                field_mask;
 
     /**
-     * Destination address; if the @ref UCP_EP_PARAMS_FLAGS_CLIENT_SERVER flag
-     * is not set, this address is mandatory for filling
-     * (along with its corresponding bit in the field_mask - @ref
-     * UCP_EP_PARAM_FIELD_REMOTE_ADDRESS) and must be obtained using
-     * @ref ucp_worker_get_address. This field cannot be changed by
+     * Destination address; one from the following fields is mandatory for
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: one *of*

File: src/ucp/api/ucp.h [hunk 15]
```diff
@@ -1664,6 +1702,21 @@ ucs_status_ptr_t ucp_ep_modify_nb(ucp_ep_h ep, const ucp_ep_params_t *params);
...
+ * @brief Reject @ref ucp_ep_h "endpoint" creation.
+ *
+ * This routine releases the @ref ucp_ep_address_h and notifies client if there
+ * was set @ref ucp_ep_params_t::err_handler with status @ref UCS_ERR_REJECTED.
+ *
+ * @param [in]  listener    Handle to the listener which created the address.
+ * @param [in]  ep_addr     Handle to the remote endpoint address to reject.
+ *
+ */
+void ucp_listener_reject(ucp_listener_h listener, ucp_ep_address_h ep_addr);
```
Comment (yosefe) [inline src/ucp/api/ucp.h]: "address" terminology implies for a stateless blob of data, but here we reject it.
maybe should change "address" terminology to "connection request"?

File: src/ucp/api/ucp_def.h [hunk 16]
```diff
@@ -81,6 +81,17 @@ typedef struct ucp_config                ucp_config_t;
 typedef struct ucp_ep                    *ucp_ep_h;
 
 
+/**
+ * @ingroup UCP_ENDPOINT
+ * @brief UCP endpoint address
+ *
+ * The address handle is an opaque object that is used as an identifier for an
+ * @ref ucp_ep_h "endpoint" instance directly connect to with
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: wording not clear

File: src/ucp/api/ucp_def.h [hunk 17]
```diff
@@ -361,6 +389,29 @@ typedef struct ucp_listener_accept_handler {
...
+/**
+ * @ingroup UCP_WORKER
+ * @brief UCP callback to handle the connection request in a client-server
+ * connection establishment flow.
+ *
+ * This structure is used for handling an incoming connection request on
+ * the listener. Setting this type of handler allows to create an endpoint on
+ * other worker and not limited to the worker associated with the listener.
+ * Other than communication progress routines, it is allowed to call other
+ * communication routines from the callback in the struct. The callback should
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h]: "should be" -> "is"

File: src/ucp/core/ucp_ep.h [hunk 18]
```diff
@@ -299,8 +296,30 @@ typedef struct {
         ucs_queue_head_t          match_q;       /* Queue of receive data or requests,
                                                     depends on UCP_EP_FLAG_STREAM_HAS_DATA */
     } stream;
+
+    struct {
+        uintptr_t                 dest_ep_ptr;   /* Remote EP pointer */
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h]: why is this moved to ext_proto? it means we would have to create ext_proto for hpc too

File: src/ucp/core/ucp_ep.c [hunk 19]
```diff
@@ -481,20 +552,20 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker, const ucp_ep_params_t *params,
     flags = UCP_PARAM_VALUE(EP, params, flags, FLAGS, 0);
     if (flags & UCP_EP_PARAMS_FLAGS_CLIENT_SERVER) {
         status = ucp_ep_create_to_sock_addr(worker, params, &ep);
-        if (status != UCS_OK) {
-            goto out;
-        }
-    } else {
+    } else if (params->field_mask & UCP_EP_PARAM_FIELD_EP_ADDR) {
+        status = ucp_ep_create_api_to_ep_addr(worker, params, &ep);
+        ucs_free(params->ep_addr);
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: are we releasing the address even if the function failed?

File: src/ucp/core/ucp_ep.c [hunk 20]
```diff
@@ -378,6 +380,75 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
...
+    }
+
+    if (status != UCS_OK) {
+        uct_iface_reject(ep_addr->listener->wiface.iface, ep_addr->conn_req_id);
+        return status;
+    }
+
+    uct_iface_accept(ep_addr->listener->wiface.iface, ep_addr->conn_req_id);
+
+    if (!((*ep_p)->flags & UCP_EP_FLAG_LISTENER)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: need a local variable `ucp_ep_h ep`

File: src/ucp/core/ucp_ep.c [hunk 21]
```diff
@@ -378,6 +380,75 @@ static ucs_status_t ucp_ep_create_to_sock_addr(ucp_worker_h worker,
...
+    return status;
+}
+
+static ucs_status_t
+ucp_ep_create_api_to_ep_addr(ucp_worker_h worker, const ucp_ep_params_t *params,
+                             ucp_ep_h *ep_p)
+{
+    ucp_ep_address_h ep_addr = params->ep_addr;
+    /* coverity[overrun-buffer-val] */
+    ucs_status_t     status  = ucp_ep_create_accept(worker, &ep_addr->priv_addr,
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: no such complex calls in variable initializer

File: src/ucp/core/ucp_listener.c [hunk 22]
```diff
@@ -90,83 +104,101 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
...
+    UCS_ASYNC_BLOCK(&worker->async);
 
-    } else {
-        status = ucp_ep_create_sockaddr_aux(listener->wiface.worker,
-                                            &params, &remote_address,
-                                            &ep);
-        if (status != UCS_OK) {
-            goto err_free_address;
+    if (listener->cb) {
+        status = ucp_listener_ep_create(listener, client_data, accept);
```
Comment (yosefe) [inline src/ucp/core/ucp_listener.c]: ucp_listener_ep_create() wrapper seems redundant, can unite the common code in this function

File: src/ucp/core/ucp_listener.h [hunk 23]
```diff
@@ -15,10 +15,14 @@
...
  */
 typedef struct ucp_listener {
-    ucp_worker_iface_t              wiface;  /* UCT iface to listen on */
-    ucp_listener_accept_callback_t  cb;      /* Listen accept callback */
-    void                            *arg;    /* User's arg for the accept callback */
-    uct_worker_cb_id_t              prog_id; /* Slow-path callback */
+    ucp_worker_iface_t                  wiface;  /* UCT iface to listen on */
+    ucp_listener_accept_callback_t      cb;      /* Listen accept callback which
+                                                    creates an endpoint */
+    ucp_listener_accept_addr_callback_t addr_cb; /* Listen accept callback which
```
Comment (yosefe) [inline src/ucp/core/ucp_listener.h]: can it be union with "cb"?

File: src/uct/api/tl.h [hunk 24]
```diff
@@ -168,6 +168,10 @@ typedef struct uct_iface_ops {
                                      const uct_device_addr_t *dev_addr,
                                      const uct_ep_addr_t *ep_addr);
 
+    void (*iface_accept)(uct_iface_h iface, void *id);
```
Comment (yosefe) [inline src/uct/api/tl.h]: pls add some opaque type for "id"

File: src/uct/api/uct_def.h [hunk 25]
```diff
@@ -289,12 +291,16 @@ typedef void (*uct_unpack_callback_t)(void *arg, const void *data, size_t length
  *
  * @retval UCS_OK         - the server will accept the connection request from
  *                          the client.
+ * @retval UCS_INPROGRESS - the user should accept or reject the request with
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: since this is not fast path, maybe easier that user could call accept/reject from the callback? to simplify the UCT API usage

File: src/uct/ib/rdmacm/rdmacm_iface.c [hunk 26]
```diff
@@ -56,12 +56,52 @@ static ucs_status_t uct_rdmacm_iface_get_address(uct_iface_h tl_iface, uct_iface
...
+        ucs_error("rdma_accept(to id=%p) failed: %m.", id);
+        rdma_reject(id, NULL, 0);
+    }
+}
+
+static void uct_rdmacm_iface_accept(uct_iface_h tl_iface, void *id)
+{
+    struct rdma_cm_event *event = id;
+
+    ucs_debug("accepting event %p with id %p", event, event->id);
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_iface.c]: ucs_trace for all logs which are per conneciton

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7212/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
also - maybe we should deprecate ucp_ep_modify_nb() and ucp_listener_accept_handler_t?
so we will have single recommended way to accept connections

File: src/ucp/core/ucp_listener.h (hunk 23 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_listener.h]: it will require other flags field and extra logic to determine which callback is set. I would prefer to have 2 pointers with simple check against NULL value until we need something else.

File: src/uct/api/uct_def.h (hunk 25 above)
Comment (evgeny-leksikov) [reply src/uct/api/uct_def.h]: we need to keep `id` untill app calls `ucp_ep_create` or `ucp_listener_reject`

File: src/ucp/core/ucp_ep.h (hunk 18 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.h]: because `listener` and `user_data` from generic extension can co-exist now, they were in union before. ep and generic ext are full with cache line size.

File: src/ucp/core/ucp_ep.c (hunk 19 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c]: what is an other option to avoid memory leak? call "reject" by user in case of failure?

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c]: if this is a transport level error then reject looks not completely working too...

File: src/ucp/core/ucp_ep.h (hunk 18 above)
Comment (yosefe) [reply src/ucp/core/ucp_ep.h]: need to refactor this so hpc use case will not use more memory

File: src/ucp/core/ucp_ep.c (hunk 19 above)
Comment (yosefe) [reply src/ucp/core/ucp_ep.c]: maybe rename "reject" to "abort"?

File: src/ucp/core/ucp_listener.h (hunk 23 above)
Comment (yosefe) [reply src/ucp/core/ucp_listener.h]: so rename the first callback field to accept_cb, to have consistent naming

File: src/uct/api/uct_def.h (hunk 25 above)
Comment (yosefe) [reply src/uct/api/uct_def.h]: yes, i mean the callback will return void and use ep_create/reject

Comment (evgeny-leksikov) [reply src/uct/api/uct_def.h]: then user must call accept or reject from callback or later but it's always his responsibility, right?

Comment (yosefe) [reply src/uct/api/uct_def.h]: yes

Comment (evgeny-leksikov) [reply src/uct/api/uct_def.h]: ok

File: src/ucp/core/ucp_ep.c (hunk 19 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.c]: "abort" sounds rough.. maybe add 3-rd function "ignore"?
really i don't see the reason why we can't release the handle here. if ucp can't create ep, the handle is useless.

File: src/ucp/core/ucp_ep.h (hunk 18 above)
Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.h]: No way without on demand allocation per endpoint (this is why we added EP allocator) or backward compatibility break (we can decide which extensions to use only at ucp_worker creation, for example, if we document that listener can be created only if wakeup feature (or some other) was requested).

Comment (yosefe) [reply src/ucp/core/ucp_ep.h]: so need to move other fields around, like user_data or err_cb

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.h]: they don't require special feature and can be requested per EP, so we must create the extension anyway.

Comment (yosefe) [reply src/ucp/core/ucp_ep.h]: ok, so pls find other way to avoid memory size increase

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.h]: OK, will add malloc/free for on demand fields.

Comment (evgeny-leksikov) [reply src/ucp/core/ucp_ep.h]: i would add "non fast path" malloc'ed extension with listener and err_cb (both not used in HPC, user_data is used on fast path in stream), we need to move at least 2 pointers to have a space for this.
union of listener with ep_match should work from first view but it looks not safety for me.

Comment (yosefe) [reply src/ucp/core/ucp_ep.h]: or modify strided_alloc to be have blocks for different stride counts
so would not have to define up front how many strides are extensions per ep

**Comment by evgeny-leksikov** [comment]
@yosefe +1 to `ucp_listener_accept_handler_t` deprecation but `ucp_ep_modify_nb` may be used in other scenario, e.g. just to reconfigure an endpoint.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7235/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@evgeny-leksikov currently if there is no use for modify_nb() let's deprecate it. we can revive it if needed. i prefer to not have redundant functions in the API.
also, jenkins failure seems relevant:
```
19:25:39 [     INFO ] Testing 65.65.65.12:45770 Interface: enP1p0s2
19:25:39 /scrap/jenkins/workspace/hpc-ucx-pr/label/r-vmb-ppc-jenkins/worker/3/contrib/../test/gtest/uct/ib/test_sockaddr.cc:148: Failure
19:25:39 Value of: delayed_conn_reqs.size()
19:25:39   Actual: 0
19:25:39 Expected: 1ul
19:25:39 Which is: 1
21:22:35 
make: *** [test] Terminated
21:22:35 
Build step 'Execute shell' marked build as failure
21:22:35 TAP Reports Processing: START
```

**Comment by evgeny-leksikov** [comment]
@yosefe ok, let's deprecate it too.
Yes, I have a fix for the failure, i'm going to update soon together with other things.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7246/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7499/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
@yosefe pls review
